luaui/Widgets/unit_factory_quota.lua

```diff
@@ -31,141 +32,138 @@ for unitDefID, uDef in pairs(UnitDefs) do
     if uDef.isFactory then
         factoryDefIDs[unitDefID] = true
         for _, opt in pairs(uDef.buildOptions) do
-            possibleFacts[opt] = possibleFacts[opt] or {}
-            possibleFacts[opt][unitDefID] = true
+            possibleFactories[opt] = possibleFactories[opt] or {}
+            possibleFactories[opt][unitDefID] = true
         end
     end
 end
 
------ Speeeed ------
+----- Speed ups ------
 local myTeam = Spring.GetMyTeamID()
-local GiveOrderToUnit = Spring.GiveOrderToUnit
-local GetFactoryCommands = Spring.GetFactoryCommands
-local GetUnitDefID = Spring.GetUnitDefID
-local GetRealBuildQueue = Spring.GetRealBuildQueue
-local GetUnitCmdDescs = Spring.GetUnitCmdDescs
-local FindUnitCmdDesc = Spring.FindUnitCmdDesc
+local spGiveOrderToUnit = Spring.GiveOrderToUnit
+local spGetFactoryCommands = Spring.GetFactoryCommands
+local spGetUnitDefID = Spring.GetUnitDefID
+local spGetRealBuildQueue = Spring.GetRealBuildQueue
+local spGetUnitCmdDescs = Spring.GetUnitCmdDescs
+local spFindUnitCmdDesc = Spring.FindUnitCmdDesc
 -----
 
-
-
-
 --------- quota logic -------------
----@param table table
----@param f function|nil
-local function findMin(table, f)
-    f = f or function(i1, i2, _, _) return i1 < i2 end
-    local best
-    local bestKey
-    for k, v in pairs(table) do
-        if not best then
-            best = v
-            bestKey = k
-        elseif f(v, best, k, bestKey) then
-            best = v
-            bestKey = k
-        end
+local function getNumberOfUnits(factoryID, unitDefID)
+    local numberOfUnits
+    if builtUnits[factoryID] and builtUnits[factoryID][unitDefID] then
+        numberOfUnits = table.count(builtUnits[factoryID][unitDefID])
+    else
+        numberOfUnits = 0
     end
-    return best, bestKey
+    return numberOfUnits
 end
 
----length of all tables
----@param table table
----@return integer
-function table.length(table)
-    local x = 0
-    for _ in pairs(table or {}) do
-        x = x + 1
+local function getMostNeedQuota(quota, factoryID)
+    local minimumQuota
+    local minimumUnitDefID
+    local minimumRatio
+    for unitDefID, quotaNumber in pairs(quota) do
+        if not minimumQuota then
+            minimumQuota = quotaNumber
+            minimumUnitDefID = unitDefID
+            minimumRatio = getNumberOfUnits(factoryID, unitDefID)/quotaNumber
+            
+        else
+            local currentRatio = getNumberOfUnits(factoryID, unitDefID)/quotaNumber
```

Comment:
Is `quotaNumber` guaranteed to never be 0?

Reply:
Yes, they are filtered out in lines 158 - 162